plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -226,6 +227,11 @@ InputStream readBlob(String blobName, long position, long length) throws IOExcep
         }
     }
 
+    public void writeBlob(String blobName, BytesReference bytes, boolean failIfAlreadyExists) throws IOException {
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucketName, blobName).build();
+        executePutMultipartBlob(blobInfo, failIfAlreadyExists, BytesReference.toBytes(bytes));
```

Comment:
`BytesReference.toBytes(bytes)` just copies all the bytes in most cases when working from anything that I'm afraid. Maybe we could use `com.google.cloud.storage.Storage#createFrom(com.google.cloud.storage.BlobInfo, java.io.InputStream, int, com.google.cloud.storage.Storage.BlobWriteOption...)` instead and use the stream we can get out of the `BytesReference` which won't do any copying?

Reply:
Sadly `Storage#createFrom` perform multiples copies 🤦. The alternative, `Storage#writer(com.google.cloud.storage.BlobInfo, com.google.cloud.storage.Storage.BlobWriteOption...)` performs multiple copies too. I'm not sure if we should apply the optimization for GCP?

Reply:
Urgh yea you're right GCP does crazy stuff there either way it seems. I'd say let's just fix Azure then and think about GCP at low priority in a follow up :)

Reply:
I'll revert the changes for GCP 👍